### PR TITLE
Using custom exception class for network loading error

### PR DESCRIPTION
**Pull request by matanshukry**
## Description

Created a custom exception class for use.

## Related Issues

#35333 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

p.s.
The implementation of the `exception.toString` method will output the message directly so that this won't be a breaking change for anyone who is dependent on the syntax of the message. Let me know if this is the right way to go.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (chunhtai) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: It feel weird to have Exception defined here, and given the information you need, it can be more general for http usage maybe something like HttpException. and I think it should be implemented in [dart sdk](https://github.com/dart-lang/sdk/blob/bbb027aa2ac4ebbb0023227347c3e2d33661b80e/sdk/lib/core/exceptions.dart#L20)
@dnfield do you have any input on this?

Comment (dnfield) [reply packages/flutter/lib/src/painting/_network_image_io.dart]: Can we just use https://api.dartlang.org/stable/2.4.0/dart-io/HttpException-class.html?

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: Please revert all of the formatting related changes.

The Flutter repository does not use dartfmt - see our style guide for more information (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/_network_image_io.dart]: We don't get the status code in that case, but maybe we coulf file a bug upstream for that if it's important.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (matanshukry) [reply packages/flutter/lib/src/painting/_network_image_io.dart]: Fixed. Check it again?

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (matanshukry) [reply packages/flutter/lib/src/painting/_network_image_io.dart]: Agreed, created an issue there. I'll probably do the PR there as well.
https://github.com/dart-lang/sdk/issues/37423

Can we put this with label blocked until the other issue is reolved?

**Comment by goderbauer** [comment]
(PR triage): Can we close this PR for now until the upstream changes are in?

**Comment by matanshukry** [comment]
@goderbauer by upstream changes you mean the issue in Dart?

And why do you want to close it? The issue is gonna stay open anyway..

**Comment by dnfield** [comment]
@matanshukry it helps us when we're trying to keep track of what actually needs attention.  This PR isn't really actionable at the moment until the Dart issue is fixed.  We should leave the issue open for sure, but if we can close this in the mean time it will help us make sure we're paying attention to pull requests that are actionable.

**Comment by dnfield** [comment]
And of course, once the upstream issue is resolved we can re-open this

**Comment by matanshukry** [comment]
@dnfield k I guess. What about the failed checks? I couldn't really figure out if it was something with my code I should fix/do?

**Comment by dnfield** [comment]
Almost certainly a flake.  It would likely pass on a rerun.

We can rerun them again when we reopen this.

**Comment by matanshukry** [comment]
@dnfield Dart team doesn't like the idea of using status code. Look up https://github.com/dart-lang/sdk/issues/37423 , latest response.

What do you recommend then? creating an HttpException in flutter (with status code), or using the NetworkImageLoadException?

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: Nits: remove blank line, and make the first line a complete sentence ending with a `.`.  Perhaps something like `The exception thrown when the HTTP request to load a network image fails.`

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: This doesn't need to be public.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: nit: end with `.`.

We should document here that `uri` must not be null.  Should also either remove `@required` from `code`, or `assert(code != null)`.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: This should describe the relationship of this status code to the request, and also if it is nullable.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: Call this `statusCode`

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: Add anewline at EOF.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: This should fail analysis because `code` is `@required`. 

But we already know that `response` is not null, so this should just be `response.statusCode` (and I realize this is a problem in the existing code :).

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/_network_image_io.dart]: I would go with `assert(code != null)`

**Comment by jonahwilliams** [comment]
This needs a test that asserts that the exception is thrown correctly, and contains the expected toString for a particular request and statusCode.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: Perhaps `The HTTP status code from the server.`

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/_network_image_io.dart]: nit: Please end files with a blank line

**Review by dnfield — CHANGES_REQUESTED** [review]
This will need tests.  For example, you could write a test that mocks out the request/response to return an error code, and validates that the correct exception is thrown with the expected status code.

**Comment by matanshukry** [comment]
I made all the changes except the tests. I'll probably create a test tomorrow. Let me know about the other changes for now. There were a lot of comments (Which is awesome btw!), but I'm pretty sure I didn't miss anything.

File: packages/flutter/lib/src/painting/_network_image_io.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/_network_image_io.dart]: This one may still need attention - and now `[code]` should be updated to `[statusCode]`.

Thanks for bearing with us on this!

**Comment by matanshukry** [comment]
Added test.

p.s.
I've added the line `export '_network_image_io.dart';` so that the exception class `NetworkImageLoadException` will be visible, although I'm not sure that's the right way to go. Let me know if there's a better way you would like me to handle it.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/image_provider.dart]: Instead of exporting here for the network image, define the exception class in image_provider.dart

Comment (matanshukry) [reply packages/flutter/lib/src/painting/image_provider.dart]: Done

**Review by dnfield — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (jonahwilliams) [reply packages/flutter/lib/src/painting/image_provider.dart]: This line needs to be removedstill

**Review by jonahwilliams — DISMISSED** [review]
Please remove the exported private library

**Comment by matanshukry** [comment]
> Please remove the exported private library

Sorry forgot about that. Fixed now.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (matanshukry) [reply packages/flutter/lib/src/painting/image_provider.dart]: Yeah forgot about it, sorry. Fixed now.

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by matanshukry** [comment]
Fro reading the guide about contributing, now that I got mr LGTM (yay :P) this PR just needs to be merged by one of the contributors with a merge access, right?

I do see 2 post-commit tests have failed, not sure if I'm suppose to do something about it though?

**Comment by dnfield** [comment]
I rekicked the test.  We can land on green

**Comment by Piinks** [comment]
Kicked `integration_tests-windows` again. If it times out again, I would recommend updating the branch.

**Pull request merged**
